net/devif/devif_send.c

```diff
@@ -68,47 +69,50 @@
 void devif_send(FAR struct net_driver_s *dev, FAR const void *buf,
                 int len, unsigned int offset)
 {
-#ifndef CONFIG_NET_IPFRAG
-  unsigned int limit = NETDEV_PKTSIZE(dev) -
-                       NET_LL_HDRLEN(dev) - offset;
+  int ret;
 
-  if (dev == NULL || len == 0 || len > limit)
+  if (dev == NULL)
     {
-      nerr("ERROR: devif_send fail: %p, sndlen: %u, pktlen: %u\n",
-           dev, len, limit);
-      return;
+      ret = -ENODEV;
+      goto errout;
     }
-#else
-  if (dev == NULL || len == 0)
+
+  if (len == 0)
+    {
+      ret = -EINVAL;
+      goto errout;
+    }
+
+#ifndef CONFIG_NET_IPFRAG
+  if (len > NETDEV_PKTSIZE(dev) - NET_LL_HDRLEN(dev) - offset)
     {
-      nerr("ERROR: devif_send fail: %p, sndlen: %u\n", dev, len);
-      return;
+      ret = -EMSGSIZE;
+      goto errout;
     }
 #endif
 
-  /* Copy in iob to target device buffer */
+  /* Append the send buffer after device buffer */
 
-  if (len <= iob_navail(false) * CONFIG_IOB_BUFSIZE)
+  if (len > iob_navail(false) * CONFIG_IOB_BUFSIZE &&
+      netdev_iob_prepare(dev, false, 0) != OK)
     {
-      /* Prepare device buffer before poll callback */
+      ret = -ENOMEM;
+      goto errout;
+    }
 
-      if (netdev_iob_prepare(dev, false, 0) != OK)
-        {
-          return;
-        }
+  /* Prepare device buffer before poll callback */
 
-      iob_update_pktlen(dev->d_iob, offset);
+  iob_update_pktlen(dev->d_iob, offset);
 
-      dev->d_sndlen = iob_trycopyin(dev->d_iob, buf, len, offset, false);
-    }
-  else
+  ret = iob_trycopyin(dev->d_iob, buf, len, offset, false);
+  if (ret != len)
     {
-      dev->d_sndlen = 0;
+      netdev_iob_release(dev);
     }
 
-  if (dev->d_sndlen != len)
```

Comment:
let's return here to avoid the check(ret < 0) after errout

Reply:
Done